### ISSUE TITLE
Make it possible to create controller-runtime and client-go fake clients with the same backing ObjectTracker

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/controller-runtime#1458]
I assume given the state of https://github.com/kubernetes-sigs/controller-runtime/issues/768 it's okay to open new feature requests for the fake client.

Context:
* Currently the controller runtime fake client provides an implementation for the `client.Client` interface: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go
* There's a similar implementation for the regular client-go `kubernetes.Interface`: https://github.com/kubernetes/client-go/blob/master/kubernetes/fake/clientset_generated.go
* Both are using testing.ObjectTracker as backing store

User Story:

We want to be able to write unit tests for a CAPI controller which requires both a controller-runtime `client.Client` and a client-go `kubernetes.Interface`. We require the client-go interface as some of our code is using upstream util funcs which require it and haven't been rewritten to use the controller-runtime client (and probably will never be). It's necessary that the code we test has a consistent view across the client-go and controller-runtime clients.

Some details:
* Both clients are using testing.NewObjectTracker internally to store and track objects
* Both are not exposing and are not allowing to hand over the ObjectTracker

Implementation ideas:

*Option 1* Adjust the fake client implementations to make it possible to hand over an ObjectTracker
  1. in controller-runtime this would probably be an additional func like: 
     ```go
     func (f *ClientBuilder) WithObjectTracker(objectTracker testing.ObjectTracker) *ClientBuilder
     ```
  2. in client-go this could be a new func like:
     ```go
     NewSimpleClientsetWithObjectTracker(objectTracker testing.ObjectTracker) *ClientSet
     ```
    
     The existing `NewSimpleClientset` would then just call `NewSimpleClientsetWithObjectTracker` internally
     This would have to be discussed with client-go maintainers of course

*Option 2* Also return an implementation of `kubernetes.Interface` in the controller-runtime fake client implementation
  * I wouldn't do this as this would duplicate the maintenance effort for the client-go fake client

What do you think?

**Comment by shashankram** [comment]
I am pretty much looking for the same thing, as we have a lot of test code leveraging client-go's fake clientset, from which informers and caches are derived. I am interested in using some of the abstractions provided by controller-runtime's cache pkg, however unit testing with it is not as flexible as with client-go's fakes.

**Comment by sbueringer** [comment]
Just fyi, has been solved in #1863.
